Apps/Inc/hashtable.h

```diff
@@ -0,0 +1,283 @@
+                    223, 76, 60, 29, 8, 173, 193, 214, 52, 122, 163, 0,
+                    113, 53, 191, 77, 80, 162, 72, 88, 231, 85, 108, 234,
+                    211, 227, 91, 205, 103, 134, 51, 243, 11, 28, 204,
+                    167, 244, 30, 97, 87, 105, 237, 141, 206, 216, 226,
+                    45, 43, 149, 236, 218, 158, 215, 61, 251, 2, 20,
+                    190, 233, 160, 95, 212, 9, 119, 169, 117, 143,
+                    202, 177, 142, 127, 92, 196, 116, 182, 98, 239,
+                    148, 70, 219, 208, 154, 188, 56, 63, 129, 179,
+                    203, 21, 35, 130, 59, 26, 153, 217, 18, 44, 69,
+                    33, 229, 39, 81, 171, 195, 15, 102, 213, 78,
+                    120, 71, 54, 176, 192, 199, 140, 3, 180, 10,
+                    93, 90, 197, 19, 253, 25, 40, 14, 112, 99,
+                    132, 101, 13, 252, 201, 22, 220, 1, 48, 146,
+                    189, 23, 68, 110, 24, 125, 240, 107, 37, 7,
+                    165, 161, 6, 128, 238, 164, 38, 135, 248, 16,
+                    32, 178, 94, 200, 247, 168, 66, 41, 5, 145, 
+                    198, 183, 89, 230, 126, 104, 75, 254, 249, 222,
+                    31, 136, 150, 114, 157, 235, 147, 83, 151, 118,
+                    138, 49, 194, 131, 74, 82, 210, 187, 96, 100, 27,
+                    64, 209, 62, 67, 172, 46, 12, 181, 186, 207, 55,
+                    245, 115};
+#endif
+
+
+//data type of hash table
+#ifndef HASH_TABLE_DATA_TYPE
+#define HASH_TABLE_DATA_TYPE int
+#endif
+
+//key type of hash table
+#ifndef HASH_TABLE_KEY_TYPE
+#define HASH_TABLE_KEY_TYPE int
+#endif
+
+//number of elements in the hashtable
+#ifndef HASH_TABLE_SIZE
+#define HASH_TABLE_SIZE 256
+#endif
+
+//name of the hashtable
+#ifndef HASH_TABLE_NAME
+#define HASH_TABLE_NAME new_hashtable
+#endif
+
+// Utility definitions
+#define _CONCAT(A, B) A ## B
+#define CONCAT(A, B) _CONCAT(A, B)
+
+//Type names
+#define HASH_ENTRY_STRUCT_NAME CONCAT(HASH_TABLE_NAME, _entry_s)
+#define HASH_ENTRY_TYPE_NAME CONCAT(HASH_TABLE_NAME, _entry_t)
+#define HASH_TABLE_STRUCT_NAME CONCAT(HASH_TABLE_NAME, _table_s)
+#define HASH_TABLE_TYPE_NAME CONCAT(HASH_TABLE_NAME, _table_t)
+
+//the struct def for the entries in the hash table
+typedef struct HASH_ENTRY_STRUCT_NAME {
+    HASH_TABLE_KEY_TYPE key;
+    HASH_TABLE_DATA_TYPE data;
+    bool inUse;
+} HASH_ENTRY_TYPE_NAME;
+
+//the struct def for the entire table
+typedef struct HASH_TABLE_STRUCT_NAME {
+    //actual data
+    HASH_ENTRY_TYPE_NAME table[HASH_TABLE_SIZE]; 
+    //other metadata
+    int remaining; //number of remaining elements
+} HASH_TABLE_TYPE_NAME;
+
+/**
+ * @brief Initialize a new hash table
+ * 
+ * If the type of the fifo is new_hashtable, then this function
+ * will be called new_hashtable_init()
+ * 
+ * @return an empty hashtable
+*/
+
+#define INIT CONCAT(HASH_TABLE_NAME, _init)
+static inline HASH_TABLE_TYPE_NAME INIT(){
+    HASH_TABLE_TYPE_NAME table;
```

Comment:
The table will go out of scope at the end of the function. You probably want the user to pass in a pointer to an already allocated table struct.